src/lcm/model_functions.py

```diff
@@ -10,28 +14,138 @@ def get_utility_and_feasibility_function(
     data_name,
     interpolation_options,
     is_last_period,
+):
+    # ==================================================================================
+    # Gather information on the model variables
+    # ==================================================================================
+    state_variables = model.variable_info.query("is_state").index.tolist()
+    choice_variables = model.variable_info.query("is_choice").index.tolist()
+    stochastic_variables = model.variable_info.query("is_stochastic").index.tolist()
+
+    # ==================================================================================
+    # Generate dynamic functions
+    # ==================================================================================
+    current_u_and_f = get_current_u_and_f(model)
+
+    if not is_last_period:
+        next_states = get_next_states_function(model)
+        next_weights = get_next_weights_function(model)
+
+        scalar_function_evaluator = get_function_evaluator(
+            space_info=space_info,
+            data_name=data_name,
+            interpolation_options=interpolation_options,
+            input_prefix="next_",
+            out_name="continuation_value",
+        )
+
+        multiply_weights = get_multiply_weights(stochastic_variables)
+
+        relevant_functions = [
+            current_u_and_f,
+            next_states,
+            next_weights,
+            scalar_function_evaluator,
+        ]
+    else:
+        relevant_functions = [current_u_and_f]
+
+    # ==================================================================================
+    # Update this section
+
+    arg_names = set()
+    for func in relevant_functions:
+        parameters = inspect.signature(func).parameters
+        arg_names.update(parameters.keys())
+    arg_names = list({"vf_arr", *arg_names})
+    arg_names = [arg for arg in arg_names if "next_" not in arg]
+
+    # Update this section
+    # ==================================================================================
+
+    # ==================================================================================
+    # Create the utility and feasability function
+    # ==================================================================================
+
+    @with_signature(args=arg_names)
+    def u_and_f(*args, **kwargs):
+        kwargs = dict(zip(arg_names[: len(args)], args, strict=True)) | kwargs
+
+        states = {k: v for k, v in kwargs.items() if k in state_variables}
+        choices = {k: v for k, v in kwargs.items() if k in choice_variables}
+
+        u, f = current_u_and_f(**states, **choices, params=kwargs["params"])
+
+        if is_last_period:
+            big_u = u
+
+        else:
+            _next_states = next_states(**states, **choices, params=kwargs["params"])
```

Comment:
I think we should pull out the if conditions from the generated `u_and_f` function and have two def statements instead.